### ISSUE TITLE
Deadlock in BulkProcessor close function

### ISSUE THREAD

**Issue by jherico** [elastic/elasticsearch#78871]
**Elasticsearch version** 
7.8.0 (but I'm pretty sure it can still happen on 7.15.0)

**Plugins installed**: 
none

**JVM version** 
```
openjdk 11.0.12 2021-07-20 LTS
OpenJDK Runtime Environment Corretto-11.0.12.7.1 (build 11.0.12+7-LTS)
OpenJDK 64-Bit Server VM Corretto-11.0.12.7.1 (build 11.0.12+7-LTS, mixed mode)
```

**OS version** (`uname -a` if on a Unix-like system):
`Linux ip-10-128-4-198 4.14.243-185.433.amzn2.aarch64 #1 SMP Mon Aug 9 05:56:00 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux`
An `m6gd.12xlarge` running the above OS

**Description of the problem including expected versus actual behavior**:
`BulkProcessor::awaitClose` can trigger a deadlock under certain conditions.

**Steps to reproduce**:

My production execution environment is a Flink job, but I believe I can describe how this could occur outside the context of Flink.

1. Start a JVM running with limited `-XX:MaxDirectMemorySize` option
1. Create a `BulkProcessor` using a `BulkProcessor.Builder` and ensure that you call `setConcurrentRequests(0)` on the builder
1. Send documents for indexing to the bulk processor inside a try block until you trigger an `java.lang.OutOfMemoryError: Direct buffer memory` exception.  This will _kill_ the nio.reactor thread
1. Inside the catch block for the above section, call `BulkProcessor.close()`

The final step will call `BulkProcessor::awaitClose()` which will then call `BulkRequestHandler::execute()` if there are pending actions.  But since the nio reactor thread is dead the `retry.withBackoff` call [here](https://github.com/elastic/elasticsearch/blob/v7.15.0/server/src/main/java/org/elasticsearch/action/bulk/BulkRequestHandler.java#L51) will never execute the passed handlers, and the latch `await` on line 67 will wait forever, even though the calling code to close specified a timeout.

I've created a small reproduction case here:  https://github.com/jherico/bug-repro/tree/es-78871  The behavior is not 100% stable, but that's not unusual for a deadlock.  However, if I run it with no breakpoints set I typically see a deadlock right at startup.

**Provide logs (if relevant)**:

Initial Exception stack trace from logs:
```
2021-10-07 12:29:42,840 ERROR org.apache.http.impl.nio.client.InternalHttpAsyncClient      [] - I/O reactor terminated abnormally
org.apache.http.nio.reactor.IOReactorException: I/O dispatch worker terminated abnormally
        at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor.execute(AbstractMultiworkerIOReactor.java:359) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager.execute(PoolingNHttpClientConnectionManager.java:221) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.client.CloseableHttpAsyncClientBase$1.run(CloseableHttpAsyncClientBase.java:64) [blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.lang.OutOfMemoryError: Direct buffer memory
        at java.nio.Bits.reserveMemory(Bits.java:175) ~[?:?]
        at java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:118) ~[?:?]
        at java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:317) ~[?:?]
        at sun.nio.ch.Util.getTemporaryDirectBuffer(Util.java:242) ~[?:?]
        at sun.nio.ch.IOUtil.write(IOUtil.java:71) ~[?:?]
        at sun.nio.ch.IOUtil.write(IOUtil.java:50) ~[?:?]
        at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:462) ~[?:?]
        at org.apache.http.impl.nio.codecs.AbstractContentEncoder.doWriteChunk(AbstractContentEncoder.java:171) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.codecs.AbstractContentEncoder.doWriteChunk(AbstractContentEncoder.java:164) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.codecs.AbstractContentEncoder.writeToChannel(AbstractContentEncoder.java:138) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.codecs.LengthDelimitedEncoder.write(LengthDelimitedEncoder.java:130) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.nio.entity.NByteArrayEntity.produceContent(NByteArrayEntity.java:136) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.nio.protocol.BasicAsyncRequestProducer.produceContent(BasicAsyncRequestProducer.java:125) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.client.MainClientExec.produceContent(MainClientExec.java:262) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.produceContent(DefaultClientExchangeHandlerImpl.java:140) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.outputReady(HttpAsyncRequestExecutor.java:249) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.DefaultNHttpClientConnection.produceOutput(DefaultNHttpClientConnection.java:290) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.client.InternalIODispatch.onOutputReady(InternalIODispatch.java:86) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.client.InternalIODispatch.onOutputReady(InternalIODispatch.java:39) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.reactor.AbstractIODispatch.outputReady(AbstractIODispatch.java:145) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.reactor.BaseIOReactor.writable(BaseIOReactor.java:187) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:341) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:591) ~[blob_p-eb802c23012999b0af2bdf1e4c964a8b6cffdf80-5873b5c6fee74d591379efbb5835118b:1.0.0]
        ... 1 more
```

Stack trace of deadlocked thread in a Flink job that basically holds up the entire system: 
```
"Map -> Sink: Unnamed (1/1)#0" #453 prio=5 os_prio=0 cpu=2222762.96ms elapsed=201025.61s tid=0x0000ffffb770f000 nid=0xd939 waiting on condition  [0x0000fff2bf7f1000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@11.0.12/Native Method)
        - parking to wait for  <0x0000fff575017580> (a java.util.concurrent.CountDownLatch$Sync)
        at java.util.concurrent.locks.LockSupport.park(java.base@11.0.12/LockSupport.java:194)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(java.base@11.0.12/AbstractQueuedSynchronizer.java:885)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(java.base@11.0.12/AbstractQueuedSynchronizer.java:1039)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(java.base@11.0.12/AbstractQueuedSynchronizer.java:1345)
        at java.util.concurrent.CountDownLatch.await(java.base@11.0.12/CountDownLatch.java:232)
        at org.elasticsearch.action.bulk.BulkRequestHandler.execute(BulkRequestHandler.java:78)
        at org.elasticsearch.action.bulk.BulkProcessor.execute(BulkProcessor.java:454)
        at org.elasticsearch.action.bulk.BulkProcessor.execute(BulkProcessor.java:463)
        at org.elasticsearch.action.bulk.BulkProcessor.awaitClose(BulkProcessor.java:330)
        at org.elasticsearch.action.bulk.BulkProcessor.close(BulkProcessor.java:300)
        at org.apache.flink.streaming.connectors.elasticsearch.ElasticsearchSinkBase.close(ElasticsearchSinkBase.java:354)
        at org.apache.flink.api.common.functions.util.FunctionUtils.closeFunction(FunctionUtils.java:41)
        at org.apache.flink.streaming.api.operators.AbstractUdfStreamOperator.close(AbstractUdfStreamOperator.java:114)
        at org.apache.flink.streaming.runtime.tasks.StreamOperatorWrapper.close(StreamOperatorWrapper.java:141)
        at org.apache.flink.streaming.runtime.tasks.RegularOperatorChain.closeAllOperators(RegularOperatorChain.java:127)
        at org.apache.flink.streaming.runtime.tasks.StreamTask.closeAllOperators(StreamTask.java:1035)
        at org.apache.flink.streaming.runtime.tasks.StreamTask$$Lambda$10378/0x0000fff2cf88cc40.run(Unknown Source)
        at org.apache.flink.streaming.runtime.tasks.StreamTask.runAndSuppressThrowable(StreamTask.java:1021)
        at org.apache.flink.streaming.runtime.tasks.StreamTask.cleanUp(StreamTask.java:928)
        at org.apache.flink.runtime.taskmanager.Task.lambda$restoreAndInvoke$0(Task.java:940)
        at org.apache.flink.runtime.taskmanager.Task$$Lambda$10371/0x0000fff2cfa89cc8.run(Unknown Source)
        at org.apache.flink.runtime.taskmanager.Task.runWithSystemExitMonitoring(Task.java:958)
        at org.apache.flink.runtime.taskmanager.Task.restoreAndInvoke(Task.java:940)
        at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:766)
        at org.apache.flink.runtime.taskmanager.Task.run(Task.java:575)
        at java.lang.Thread.run(java.base@11.0.12/Thread.java:829)
```

**Comment by jherico** [comment]
I've created a bit of code to try to reproduce this deadlock here:  https://github.com/jherico/bug-repro/tree/es-78871

**Comment by original-brownbear** [comment]
This one is pretty serious now. With the improvements to ILM which trigger a much higher rate of ILM task completion than previously possible, this reproduces a deadlock in our many shards benchmarking reliably.

**Comment by masseyke** [comment]
I think this is the same as #68468 right? Thanks for the test code -- I'll see if it passes with the changes I'm making.

**Comment by original-brownbear** [comment]
This is the cause of #68468 I'd say. The solution to the fact that the bulk processor deadlocks is different from solving #68468 potentially, since for the general use-case it might be ok to introduce a solution that eventually blocks writes if the queue of unwritten data gets too long (which I believe is the currently expected behavior).

**Comment by mattc58** [comment]
@masseyke do the recent changes to BulkProcessor handle this, and can we close this one?

**Comment by masseyke** [comment]
I don't believe we've done anything with BulkProcessor, so this is still possible in 7.x when client code uses it directly like this. However all internal uses of BulkProcessor have been replaced with (the temporarily-named) BulkProcessor2, which avoids these kinds of deadlock problems. And as of 8.0 or so, BulkProcessor is no longer available to client code -- it's internal to Elasticsearch. Users are expected to use the new Java API. From what I can tell, there is no Flink connector for Elasticsearch 8 yet (https://issues.apache.org/jira/browse/FLINK-26088).
So to summarize:

- This is still a problem for customers using Elasticsearch 7.17, including Flink connector users
- This is not a problem for customers using Elasticsearch 8.x

**Comment by dakrone** [comment]
Okay, with that info, I think we should close this for now, and we can revisit it if we ever intend to address it in 7.17.x (which seems unlikely). Closing for now and we can re-open if needed.
